packages/core/src/intl/languages.ts

```diff
@@ -149,107 +117,137 @@ export function getNotifications<
+) {
+    const candidates = languages.filter((lang) => Locale.isLanguageCode(lang.code, langCode));
+
+    if (candidates.length === 1) {
+        return langCode;
     }
 
-    // Try shorting to region code
-    const shortRegionCode = locale.toRegionCode();
-    const matchingLanguagesByRegionCode = languages.filter(
-        ({ code }) => LocaleObject.fromAnyCode(code).toRegionCode() === shortRegionCode,
-    );
+    return undefined;
+}
+
+function getUnambiguousRegionCode<Language extends Pick<NewsroomLanguageSettings, 'code'>>(
+    languages: Language[],
+    regionCode: Locale.RegionCode | undefined,
+) {
+    if (!regionCode) {
+        return undefined;
+    }
+
+    if (isNumberCode(regionCode)) {
+        // We don't want just numbers in our region code
+        return undefined;
+    }
+
+    const candidates = languages.filter((lang) => Locale.isRegionCode(lang.code, regionCode));
     // Prevent collision with neutral language codes
-    const lowerCasedShortRegionCode = shortRegionCode.toLowerCase();
-    const matchingNeutralLanguagesByRegionCode = languages.filter(
-        ({ code }) =>
-            LocaleObject.fromAnyCode(code).toNeutralLanguageCode() === lowerCasedShortRegionCode ||
-            code === lowerCasedShortRegionCode,
+    const collisions = languages.filter((lang) =>
+        Locale.isLanguageCode(lang.code, regionCode.toLowerCase()),
     );
-    if (
-        matchingLanguagesByRegionCode.length === 1 &&
-        // If there are 2 or more matching neutral languages, it means that there are no languages that can be shortened to neutral code
-        matchingNeutralLanguagesByRegionCode.length !== 1 &&
-        // We don't want just numbers in our region code
-        Number.isNaN(Number(shortRegionCode))
-    ) {
-        return shortRegionCode;
+
+    if (candidates.length === 1 && collisions.length === 0) {
+        return regionCode;
     }
 
-    // Return the original (exact) code if shortening is not possible
-    return localeCode;
+    return undefined;
 }
 
 /**
- * Get matching language for the requested locale
- * The logic is reversed from `getShortestLocaleCode`, so that it "unwraps" the possible variants with no collisions
- *
- * First: get by exact match
- * Then: get by matching region code part
- * Finally: get by matching language code part
+ * Get matching language for the requested locale slug.
+ * The logic is reversed from `getShortestLocaleSlug`, so that it "unwraps" the possible variants with no collisions
  *
- * @param locale A LocaleObject constructed from the requested locale code (taken straight from the URL)
+ * The order of checks is:
+ * - exact locale slug match
+ * - match language code, preferring non-default languages
+ * - match region code, preferring non-default languages
+ * - match language code across all languages
+ * - match region code across all languages
  */
-export function getLanguageFromLocaleIsoCode<
+export function matchLanguageByLocaleSlug<
```

Comment:
> We're now checking the default language (which shouldn't have a locale slug)

I don't see anything special about the default language. Even if it's not expected to have a locale slug in the path, it can still be accessed with it. If that happens the URL location will be considered non-canonical and the redirect will be triggered.

_Scenario_:

- _Enabled languages: `en-US` (default), `en-CA`, `fr-CA`_
- _Requested URL_: `/us`

It should still redirect to the `/` homepage. But if the default language was excluded, this would produce a 404. Right?

Reply:
> Full locale slugs are now also checked for partial match

Hm. I don't understand this sentence.

How are slugs checked for a partial match? I believe they are not.

> UPD: I can see below that `matchLanguageByLangSlug` will return only if there's only one candidate. But it's not really obvious from the function name, so maybe it's worth adding some extra comments for future generations :)

Hm... 

This comment makes me think that we should remove the single-candidate requirement. This is different from URL generation (`getShortestLocaleSlug()`)  where we have to ensure the generated slug uniquely identifies a locale without collisions. In this scenario, we should be more permissive and in case of collisions it should still return one of the enabled locales (and redirect to the canonical URL). 

Right?

Reply:
> Full locale slugs are now also checked for partial match

I described the scenario in `en-GB`, but not sure if that's what would actually happen with new logic.

Before, when you were requesting a full locale slug (eg `en-GB`) it would only check for exact match, since if there's no exact match, there's no possible match for this locale slug. From the new code, it looks like you it would be matched with other English languages or GB region, but I don't think this is correct behavior (user requesting this specific locale is probably not expecting another language or region)

Reply:
> From the new code, it looks like you it would be matched with other English languages or GB region, but I don't think this is correct behavior (user requesting this specific locale is probably not expecting another language or region)

No, this is not what happens.

- `matchLanguageByLangSlug()` is only checking if language lang code is matching the full request slug. 
- `matchLanguageByRegionSlug()` is only checking if language region code is matching the full request slug.

In both cases, the full request slug is used. So if `en-gb` is requested, the only language that can be matched with it is `en_GB`.

I've converted the function with tests in ed0c2f72f22c09ed4e0e337fb5b39485f5a0bf21, and also tweaked its behaviour to be more permissive in case of collisions as described [in my comment above](https://github.com/prezly/theme-kit-js/pull/403#discussion_r1366788110).

Reply:
Good, thanks for the explanation! 👍🏻